mlir/lib/Dialect/Vector/IR/VectorOps.cpp

```diff
@@ -2047,6 +2047,49 @@ static Attribute foldPoisonSrcExtractOp(Attribute srcAttr) {
   return {};
 }
 
+static Attribute foldDenseElementsAttrSrcExtractOp(ExtractOp extractOp,
+                                                   Attribute srcAttr) {
+  auto denseAttr = dyn_cast_if_present<DenseElementsAttr>(srcAttr);
+  if (!denseAttr) {
+    return {};
+  }
+
+  if (denseAttr.isSplat()) {
+    Attribute newAttr = denseAttr.getSplatValue<Attribute>();
+    if (auto vecDstType = llvm::dyn_cast<VectorType>(extractOp.getType()))
+      newAttr = DenseElementsAttr::get(vecDstType, newAttr);
+    return newAttr;
+  }
+
+  auto vecTy = llvm::cast<VectorType>(extractOp.getSourceVectorType());
+  if (vecTy.isScalable())
+    return {};
+
+  if (extractOp.hasDynamicPosition()) {
+    return {};
+  }
+
+  // Calculate the linearized position of the continuous chunk of elements to
+  // extract.
+  llvm::SmallVector<int64_t> completePositions(vecTy.getRank(), 0);
+  copy(extractOp.getStaticPosition(), completePositions.begin());
+  int64_t elemBeginPosition =
+      linearize(completePositions, computeStrides(vecTy.getShape()));
+  auto denseValuesBegin =
+      denseAttr.value_begin<TypedAttr>() + elemBeginPosition;
+
+  TypedAttr newAttr;
+  if (auto resVecTy = llvm::dyn_cast<VectorType>(extractOp.getType())) {
+    SmallVector<Attribute> elementValues(
+        denseValuesBegin, denseValuesBegin + resVecTy.getNumElements());
```

Comment:
Can this cause IR size to explode? For example, consider a large non-splat constant of size N and a series of K extracts that extract half of the elements, each at a different offset. I think this would increase IR size from N + K to K * (N / 2), wouldn't it?

Reply:
You cannot extract half the elements using vector.extract. You can only index an outer most dimension. Because of this, you can really only use N more elements worth of space, if N is the number of elements in the original dense attribute.

Reply:
Ah OK, so this would only be a concern with extract_strided_slice, right?

Could you add this explanation as a code comment to make it obvious why creating these constants won't be a problem?

Reply:
I think we can create an issue to investigate this, but I don't think it is for this patch.

Reply:
> Ah OK, so this would only be a concern with extract_strided_slice, right?
> 
> Could you add this explanation as a code comment to make it obvious why creating these constants won't be a problem?

Yeah I think it's mostly for extract_strided_slice. I can add a comment, yes.

Reply:
> For example, consider a large non-splat constant of size N and a series of K extracts that extract half of the elements, each at a different offset. I think this would increase IR size from N + K to K * (N / 2), wouldn't it?

Could you provide a specific example? I'm not sure I follow how this would increase the IR size. If we fold one of these `vector.extract` ops, either the constant attribute is used directly and no constant op is generated for it (best case, -1 `vector.extract`) or a constant op is generated for it (worse case, -1 `vector.extract`, +1 `arith.constant`). What am I missing?

Reply:
Something like this (pseudo-ir):
```
X = dense<[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]>
A = extract %X[0:5] // [0, 1, 2, 3, 4]
B = extract %X[1:6] // [1, 2, 3, 4, 5]
C = extract %X[2:7] // [2, 3, 4, 5, 6]
D = extract %X[3:8] // [3, 4, 5, 6, 7]
E = extract %X[4:9] // [4, 5, 6, 7, 8]
F = extract %X[5:10] // [5, 6, 7, 8, 9]
```
Folding this sequence would replace a dense elements attr of size 10 with 5 unique dense elements arrays of size 5 (25). But like Kunwar said, I think this only comes up in the extract_strided_slice case.

Reply:
added